### ISSUE TITLE
Proposal: shorter release cycle

### ISSUE THREAD

**Issue by morosanuae** [Jermolene/TiddlyWiki5#3834]
Hi to all,

@Jermolene why we have to wait for such a long time for a new release? For example, v5.1.20 (now in development) already contains a lot of useful things that will be available after a few months from now in the current rhythm of releasing. Wouldn't be much better if each release will ship fewer features and maybe more bug fixes that will lead to a more stable and concentrated release? That way the awesome new features will be available sooner and also the bug fixes that affects the functionality in current release. And from what I see the incompatible changes are very rare, especially with the 5.1.x minor versions. I see only pros to proposed approach. Are there any cons that stops you from doing this?
Maybe other will share their opinion. I think this is a very important aspect.

Thanks in advance.

**Comment by Jermolene** [comment]
Hi @morosanuae I'd welcome other views, but my own is that there is a tricky balance to be struck. For example, one factor is that the more frequent we make releases the less likely it is that people will upgrade to every new release.

We've had some long gaps in the past which have mostly been due to external factors like illness. I think a new release every three months or so is about right, so I guess we are certainly about due for 5.1.20...

**Comment by morosanuae** [comment]
@Jermolene Firstly I don't think is mandatory or even useful for all people to upgrade to every new release. I guess the major motivation for an upgrade should be for bug fixes in the first place. Maybe many users (including me) will not care that much about some new introduced features but the bug fixing is always of interest for many (including me). Even 3 months is a very long time when you wait for a crucial fix to be released. I expect, and maybe others too, that each release to be as stable and bug free as it can be, to be a quality release and if that's not possible for obvious reasons then the shorter release cycle can be a solution because you will obtain feedback quicker. IMHO new features can be released on a 3 months interval or even longer, but not the bug fixes.
Anyway for me at least a smaller release will be easier to digest and incorporate in my current work.
But of course human (external) factors are also involved, and if you feel overwhelmed or you don't have time/energy I understand, it's not like you are paid for this. It's just a proposal anyway...

**Comment by MidnightLightning** [comment]
I've seen more and more projects these days get less shy about creating new "major" version releases (e.g. Chrome releases), which allows shifting the mentality of what the classic X.Y.Z / Major, Minor, Patch versioning means. I'd be in favor of following the [Semantic Versioning](https://semver.org/) guideline of having patch-releases being backwards-compatible bugfixes (so, more frequent than current, but with only bugfixes), and having minor version releases be ones where new features are added (probably more in line with the current 3-month release timelines)?

**Comment by Jermolene** [comment]
Hi @morosanuae 

>  Firstly I don't think is mandatory or even useful for all people to upgrade to every new release

There's a balance to be struck. We really want as many people as possible to upgrade so that we can catch regressions. We cater for people who want to be on the bleeding edge by making the prerelease available.

Hi @MidnightLightning at the moment we get away with working with just two branches, "master" for the prerelease and "tiddlywiki-com" for the latest content for tiddlywiki.com. Having independent bug fix releases would make things more complicated: we'd need a branch for the latest feature release as well as the latest bugfix release. Again, there's a balance to be struck; more active branches increases the chances of human error, and condemns all of us to more git terrors as we try to keep them up to date with one another.

**Comment by MidnightLightning** [comment]
> we'd need a branch for the latest feature release as well as the latest bugfix release

I'm not sure I'm grasping what you mean here? Once you do a "feature release" (e.g. 5.30.0), and push it to the website as released (the "current" version), if you find a bug in it and fix it (5.30.1), wouldn't you want that to be pushed live immediately (there's no reason for people to hang out on 5.30.0 with a known bug still in it, if 5.30.1 is backwards-compatible)? So, you wouldn't need to keep a branch around for the 5.30.0 release after 5.30.1 is out, right?

From a git branching perspective, you say the current model is `master` and `tiddlywiki-com`, representing "prerelease" and "the website". In other projects I've seen that follow a ["git-flow"-like model](https://nvie.com/posts/a-successful-git-branching-model/), those fit into the `develop` and `master` ideas, and could continue in that role. So, taking for example if the `tiddlywiki-com` branch has 5.30.0 as its latest/current commit, then the `master` branch is working on the development of what will be 5.31.0 when it's released. If a new bug is reported, and it's a high-priority fix, a new branch would have to be created off of `tiddlywiki-com` (representing 5.30.1), bugfix work done there, and then when the fix is confirmed, it would get merged into the `tiddlywiki-com` branch _and_ into `master` (such that the bug is fixed in the 5.31.0 release too). So, there is a branch created for the bugfix/patch release, but the concept is that those are infrequent/rare, so it's not like you _always_ have those branches to deal with.

Though this might be more me not being as afraid of git branches from working in a software development firm for the last decade!

**Comment by morosanuae** [comment]
@MidnightLightning I guess @Jermolene doesn't have enough spare time to review and manage all of this being the only "authorized" person. Being the only "admin" is not necessarily a bad thing because this way things remain consistent. The downside (a major one IMO) is that the project is moving forward very slow. Slower than the speed of contributions arrival. I don't know if there's something we can do about it. I think this slow moving pace also has a negative impact among many contributors (including me) and I've noticed that some of them even "disappeared" from the scene which is a very bad thing for this project.

**Comment by sukima** [comment]
After seeing many OOS projects on GitHub come and go I can say that TW is on the *well-maintained* list. @Jermolene is very responsive compared to other developers. I've had PRs linger for **years** before an author responds.

I really do not get the impression that this project has atrophy. There are some legitimate issues that lead to the cause of a slower release cycle. For starters TiddlyWiki has a reputation for being backwards compatible or offering a smooth migration path. To accomplish this there has to be a high level of vigilance when it comes to releasing official code. Especially when contributions require a dialog and one or both parties ignore their GutHub email notifications.

An actual *release* is nothing more then an NPM publish and a tiddlywiki.com update. The act of your own TiddlyWikis upgrading is a step for the user. That same step is really no different then downloading the latest development version on GitHub. A snapshot of `master` is pretty easy and you have the cutting edge version. An official release is not necessary to upgrade or experiment.

That leads me to wonder is the complaint about actual releases or in the communication feedback cycle regarding pull-requests?

**Comment by Jermolene** [comment]
Hi @MidnightLightning 

> I'm not sure I'm grasping what you mean here? Once you do a "feature release" (e.g. 5.30.0), and push it to the website as released (the "current" version), if you find a bug in it and fix it (5.30.1), wouldn't you want that to be pushed live immediately (there's no reason for people to hang out on 5.30.0 with a known bug still in it, if 5.30.1 is backwards-compatible)? So, you wouldn't need to keep a branch around for the 5.30.0 release after 5.30.1 is out, right?

The need for an extra branch is if we wanted to be able to push documentation updates while 5.30.1 is under development.

**Comment by Jermolene** [comment]
Hi @sukima @morosanuae @MidnightLightning 

I may have said this before, but the surprising thing I've learned about PRs over the years is that they take several times the effort to process than would be involved if I did the work myself. Reviewing code in GitHub is a much more passive mental posture than writing it, and it's hard to get ones brain properly thinking through all the consequences of every change.

(To be fair, there are whole classes of PRs that are relatively straightforward to merge, but you tend not to see those ones hang around. Everybody focusses on the number of open PRs but it's really just as important to look at the number of PRs that are merged).

At times when I get overly worried about keeping people happy, I've many times ended up merging PRs without checking them properly, and lived to bitterly regret it. That's not a criticism of the authors of the PRs; we all make mistakes, and they have every right to expect that their work will be properly checked before merging.

So, with PRs we've got a large pile of intricate tasks where the price of getting it wrong is the primary long term existential risk to the project.

In terms of issues, there's an expectation that people can use them as discussion boards which means that they don't necessarily have a clear resolution. I don't like having issues that are not actionable, but equally it seems unfriendly to close those that aren't.

>  I think this slow moving pace also has a negative impact among many contributors (including me) and I've noticed that some of them even "disappeared" from the scene which is a very bad thing for this project.

It's very frustrating when that happens, and I'd like to avoid it. We can't solve it by being less vigilant about modifications, but there may be other things that would improve the situation.

Finally, I should acknowledge that I certainly struggle to get everything done that I'd like to with Tiddlywiki; there's not enough hours in the day, and I have to necessarily focus on the work that pays my bills. The good thing is that all of my current work is based on Tiddlywiki, so even if I'm not merging PRs all day, I am generally working on some aspect of Tiddlywiki.

**Comment by MidnightLightning** [comment]
> The surprising thing I've learned about PRs over the years is that they take several times the effort to process than would be involved if I did the work myself. Reviewing code in GitHub is a much more passive mental posture than writing it, and it's hard to get ones brain properly thinking through all the consequences of every change.

Interesting! Myself having worked in a software development firm for the last decade, I think I've gotten more used to pull requests as a review tool, and I wonder if the issues you're describing here could be mitigated with some process changes? I hear you say it's an incredibly manual process, which to me sounds like could be fixed with some automated testing, or at least a formal testing plan? I see you have in a few side branches started to experiment with Travis CI, which I haven't personally used that much, but any sort of CI process could help here (my team at my day job has moved over to a GitLab-based infrastructure, and do love its CI/CD infrastructure; if you can do it in Docker, you can automate it!). Even if it's not an automated process, having a testing checklist written out (I'm guessing the key things would be making sure all the "saver" methods still work, browsers that are claimed to be supported still work, exporting content and importing it back in with the same version works, and importing content exported with a previous version still works) could help make you more sure you didn't miss anything, and could allow the community to help by running through the checklist themselves and giving a "works for me!" vote of confidence in the merge.

> So, with PRs we've got a large pile of intricate tasks where the price of getting it wrong is the primary long term existential risk to the project.

Indeed, though that's true of _any_ software development project. Perhaps doing something like formally having "release candidates" for beta testers in the community to give feedback on would be a helpful process in trying to catch any unintended consequences of new development?

> In terms of issues, there's an expectation that people can use them as discussion boards which means that they don't necessarily have a clear resolution. I don't like having issues that are not actionable, but equally it seems unfriendly to close those that aren't.

This might be a good use-case for using GitHub "tags" on issues? That could help make it clear to people which ones are in a "gathering ideas from the community" phase, versus ones that are in a "consensus reached, waiting for development" phase. Since there is a fairly active Google group for the community, I don't personally think it's unfriendly to close issues that a general consensus is reached of "won't do this right now; possibly down the road we'll revisit".

**Comment by Jermolene** [comment]
Thanks @MidnightLightning 

We do in fact have a test suite with very patchy coverage. It lives in its own edition:

https://github.com/Jermolene/TiddlyWiki5/tree/master/editions/test/tiddlers/tests

We do also use Travis-CI. It runs the tests on all PRs, and performs the builds for tiddlywiki-com and master.

https://travis-ci.org/Jermolene/TiddlyWiki5

One thing I'd love to automate is coding style checks; at the moment it's very labour intensive to manually verify and seek changes, partly because our coding standards are fairy unusual.

It would certainly help if the tests were more complete. But I'm not sure that it's the biggest hold up. Most of the delay is lack of time on my part, but the contentious PRs tend to be held up by design issues. For example, considerations about backwards compatibility, or limiting our choices for the future. Some examples of PRs like that include #3891, #3854, #3884.

In terms of the release process, we make "master" available as the prerelease, and then when we make an announcement when get closer to the next release requesting help with the final testing.

We should definitely be making more use of tags. As you scroll back through the PR listing you'll see that I've attempted to add them in the past. One minor problem is that I might tag something as "needs revision", but I can't easily see that the revision has taken place. Perhaps there should be some tags that the owner of the PR can set/unset.

**Comment by MidnightLightning** [comment]
> We do also use Travis-CI. It runs the tests on all PRs...

You sure? It seems to only run on PRs you create; looking at the list of pull requests for the project, only the ones opened by you are the ones that have the CI checkmark/status symbol on them:

![Screen Shot 2019-04-29 at 10 00 19 AM](https://user-images.githubusercontent.com/250250/56905389-b4daf380-6a65-11e9-8469-8af4248a26ff.png)

Possibly it's not running on pull requests from forked repos?

**Comment by Jermolene** [comment]
> Possibly it's not running on pull requests from forked repos?

My mistake, that's correct.  My concern in setting it up was to ensure that a malicious PR couldn't steal my GitHub token; we're using a manual GitHub push, and so as far as I can tell have no choice but to use an environment variable to pass the token.

> moderators who have the permissions to add/remove tags could help curate the backlog (but not the rights to do a merge?), to keep things up-to-date (e.g. when a design review is done) would be a way to help manage that?

Gaining those fine grained permissions is the motivation behind #3114. The last time I checked, personal repos can only invite all-or-nothing collaborators, who have full permissions.

**Comment by MidnightLightning** [comment]
> My concern in setting it up was to ensure that a malicious PR couldn't steal my GitHub token; we're using a manual GitHub push, and so as far as I can tell have no choice but to use an environment variable to pass the token.

Huh; I don't have experience setting up TravisCI with repos, but I  know other pull requests of mine that use Travis (e.g. [this one](https://travis-ci.org/ethereum/ethereum-org/builds/273473001)) did get the CI jobs run on them, even though I wasn't a maintainer of the project. Though that might also relate to the difference between a personal repo and a "company" repo?

**Comment by Jermolene** [comment]
Hi @MidnightLightning sorry for the confusion: the restrictions are intentional, and the rationale is the issue about GitHub tokens that I mentioned. And there are indeed still a frustrating number of differences between personal repos and organisations.

**Comment by MidnightLightning** [comment]
Indeed! Maybe one more option to throw into the stew is instead of moving to a GitHub "organization" account, move over to GitLab (or some other host) that better suits the needs of the project? Or is that too big of an idea to analyze for the time being?

**Comment by Jermolene** [comment]
To be clear, we don't have a problem with moving to an organization repo, it's obviously more appropriate than the current setup. The hold-up is that we're anxious about unforeseen impacts of the move. I plan to migrate TiddlyDesktop first as a trial, but haven't had time to do that yet.

**Comment by ento** [comment]
> [Travis] restrictions are intentional, and the rationale is the issue about GitHub tokens that I mentioned

According to [Travis's documentation](https://docs.travis-ci.com/user/environment-variables#defining-variables-in-repository-settings), "[Environment] Variables defined in repository settings are [..] not automatically available to forks." It seems like it's safe to enable pull request builds then, if protection of the token was the sole concern..? I'd very much like to have automatic testing on PRs, as the lack of it contributes to #3313.

> To be clear, we don't have a problem with moving to an organization repo, it's obviously more appropriate than the current setup. The hold-up is that we're anxious about unforeseen impacts of the move.

One idea on this that can reduce the risk: assuming that most of the probable risk is from URLs under github.com/Jermolene/TiddlyWiki5 disappearing after the migration, the old repo can be left intact (perhaps [archived](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/archiving-a-github-repository)) with a conspicuous message pointing to the new organization repo. To preserve GitHub issues, it can be done in two steps:

1. Move the repo to the new organization.
2. Re-create the old repo by forking the new one or pushing the git repository again.

**Comment by Jermolene** [comment]
Hi @ento 

> According to Travis's documentation, "[Environment] Variables defined in repository settings are [..] not automatically available to forks." It seems like it's safe to enable pull request builds then, if protection of the token was the sole concern..? I'd very much like to have automatic testing on PRs, as the lack of it contributes to #3313.

Thank you, that's excellent. I've gone ahead and turned on PR builds.

> One idea on this that can reduce the risk: assuming that most of the probable risk is from URLs under github.com/Jermolene/TiddlyWiki5 disappearing after the migration, the old repo can be left intact

GitHub apparently automatically redirects from the old URL to the new one. I'm not sure of the details of how it works (e.g. can I disable the redirects by creating a new repository with the same name?). The plan is to experiment with TiddlyDesktop first.

There is actually another issue: whether we should move TiddlyWiki5 into the TiddlyWiki organisation, or create a new TiddlyWiki5 organisation. I'm thinking primarily of convenience for visitors who might be confused by material for the two versions being intermixed, but it would also likely simplify setting up permissions. Then we've got TiddlyDesktop which applies to both TW Classic and TW5. It seems overboard to create a TiddlyDesktop organization but it would give us a natural way to create an independent website for it.

**Comment by ento** [comment]
>  I've gone ahead and turned on PR builds.

Thank you! One follow-up PR: #4788

> There is actually another issue: whether we should move TiddlyWiki5 into the TiddlyWiki organisation, or create a new TiddlyWiki5 organisation

I realized there's a GitHub issue for the migration: #3114 - I'll comment over there.

**Comment by pmario** [comment]
Close this one:

- The OP is from 2019 and there are strong reasons why our release cycles are longer. 
- The test system is now provided by Netlify and seems to work
